teps/0005-tekton-oci-bundles.md

```diff
@@ -0,0 +1,481 @@
+
+
+### Goals
+
+The goal of this TEP is to define the specification for the Tekton OCI
+bundles. The goal of the Tekton OCI bundles is to define a Spec on top
+of a known, widely used transport mechanisms, which are the OCI
+artifact (OCI images, …).
+
+### Non-Goals
+
+How Tekton OCI bundles are used (referred to) in the different Tekton
+components is out of scope and will be the subject for an additional
+TEP. Some examples in the doc might suggest how we would reference
+those in Pipeline CRDs but they are only examples, not propositions.
+
+The tooling around Tekton OCI bundles is also out of scope.
+
+
+## Requirements
+
+- Images have a defined format that allows Tasks, Pipelines, and any
+  other Tekton specific types to be written in, and read out.
+- The spec should be documented well enough such that any user could
+  write tooling to generate a valid Tekton Bundle
+
+## Proposal
+
+When using a Tekton Bundle in a task or pipeline reference, the OCI artifact backing the
+bundle must adhere to the following contract.
+
+Our most important capability is enabling references of remote Tasks
+from within TaskRuns or Pipelines and Pipelines from within
+PipelineRuns.
+
+With that in mind, a little background on the OCI format is
+necessary. For a deeper understanding see
+https://github.com/opencontainers/image-spec. At its most basic an
+image is a manifest specifying a set of Layers referenced by a sha256
+digest. Each Layer has a media type and can have annotations. Looking
+at other similar projects (eg Helm 3) and building off of the
+knowledge from our experiment, we have arrived at the following spec:
+
+- An image will store each resource as a new layer
+  This allows us to quickly store and retrieve individual resources
+  because we can key the layer they reside in against its metadatax
+- Each layer will have a `org.opencontainers.image.title` annotation
+  (this a commonly used annotation:
+  [link](https://github.com/opencontainers/image-spec/blob/master/annotations.md#pre-defined-annotation-keys))
+  that contains the `ObjectMetadata.Name` field of the resource. Other
+  common annotations can be used like
+  `org.opencontainers.image.authors`, etc.
+- Each layer will have a `dev.tekton.image.kind` annotation which
+  specifies the Kind of the resource. For example Kind: Task would
+  become “task” and Kind: Pipeline would become “pipeline”.
+- Each layer will have a `dev.tekton.image.apiVersion` annotation which
+  specifies the apiVersion of the object (eg, v1beta1).
+
+This choice presents some tradeoffs:
+
+- Two objects of the same `{name, type, apiVersion}` will be
+  rejected. This is purposeful because it replicates existing
+  in-cluster functionality. You wouldn't be able to store two Tekton
+  Tasks of the same `name` and `apiVersion` in a namespace so to make
+  the task references easier to reason about, we enforce the same
+  characteristic on the images.
```

Comment:
You are right that registries will not enforce this. Likely, we would enforce this on pull. In particular because we rely on these annotations to understand what resource is in each layer, if it isn't specified, it is easy to reject the OCI image as it will be unusable.

Reply:
Makes sense! Another approach might be to treat these layers just like an actual image layers, and have the last one win...

Not sure which is better. I like the idea of not imposing extra semantics on what can be a valid bundle, but this is probably a rare case anyway.